### PR TITLE
Sketcher: Delete Equal constraints on LineSegments in SketchObject::trim

### PR DESCRIPTION
**Pull request by davidosterberg**
This closes:
https://tracker.freecadweb.org/view.php?id=4510

Forum discussion:
https://forum.freecadweb.org/viewtopic.php?f=8&t=53240

After PR:
fix

Before PR:
weird_sketcher_trim_behavior_simpliied

### CODE REVIEW

**Comment by davidosterberg** [comment]
@abdullahtahiriyo, can you help me through this one. It is the first time I look at the source of sketcher. Also not an experienced c++ coder.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -2313,14 +2334,16 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
                 std::swap(x1,x2);
             }
             if (x1 >= 0.001*length && x2 <= 0.999*length) {
-                if (x1 < x0 && x2 > x0) {
+                if (x1 < x0 && x2 > x0) {  // trim a way a sement in the middle
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:2337]:
```suggestion
                if (x1 < x0 && x2 > x0) {  // trim away a segment in the middle
```

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -2291,6 +2291,27 @@ int SketchObject::trim(int GeoId, const Base::Vector3d& point)
         return false;
 
     };
+    
+    // this is a helper function to remove Equal constraints from Line segments. 
+    auto delEqualConstraintsOnGeoId = [this] (int GeoId) {
```
Comment (chennes) [inline src/Mod/Sketcher/App/SketchObject.cpp:2296]: I might actually spell out "delete" here (or change to "remove") -- my brain read it as "delta" first.

**Comment by chennes** [comment]
I see you are only applying this to line segments: does the Equal constraint actually make sense for all the other shapes getting trimmed?

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 2 above)
Comment (davidosterberg) [reply src/Mod/Sketcher/App/SketchObject.cpp:2296]: Yes, I agree. But the convention in the file (and the method) seems to be to use the prefix _del_. I.e _delConstraint_ , _delConstraints_ and  _delConstraintOnPoint_ etc. I don't mind either way.

**Comment by davidosterberg** [comment]
> I see you are only applying this to line segments: does the Equal constraint actually make sense for all the other shapes getting trimmed?

Yes. it only makes sense on LineSegment. For arcs the Equal constraint is applied for the radius not the arc-lenth. Since the radius is invariant under trim operation it makes sense to keep the Equal constraints.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 2 above)
Comment (chennes) [reply src/Mod/Sketcher/App/SketchObject.cpp:2296]: OK, sorry, I did not look at the rest of the file. I think go with the convention, then.

**Review by chennes — APPROVED** [review]
This compiles, runs, and behaves as expected for the straight line segment cases I tested.

**Comment by davidosterberg** [comment]
Travis failed for Bionic gcc-7.
Not sure it is related to me. I cant reproduce locally because running the tests crashes Kde Plasma and FreeCAD gets killed. This was for master not my branch. 

```
FAIL: test20 (PathTests.TestPathCore.TestPathCore)
Test Path Tool and ToolTable object core functionality
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Mod/Path/PathTests/TestPathCore.py", line 155, in test20
    self.assertEqual(str(table.Tools), '{1: Tool 12.7mm Drill Bit, 2: Tool my other tool}' )
AssertionError: '{1L: Tool 12.7mm Drill Bit, 2L: Tool my other tool}' != '{1: Tool 12.7mm Drill Bit, 2: Tool my other tool}'
...

**Comment by chennes** [comment]
> Travis failed for Bionic gcc-7.
> Edit: After inspecting test20 in "/usr/local/Mod/Path/PathTests/TestPathCore.py", I have to say that it does not look Sketcher related.

Travis has been broken (due to OS X build not finishing in time) for a few weeks so we've been ignoring its failures: this one crept into a recent update to Path, I think you can be pretty confident it doesn't have anything to do with your code.

**Comment by abdullahtahiriyo** [comment]
@davidosterberg 

I see you have tried to contact me above. I am having a real problem with GitHub notifications. I am getting hundreds of them per day, to a point that I am simply ignoring them. This is bad because I miss things (like this one, or comments on reviews), but I cannot bear with so many notifications. I am trying different settings, but I have not arrived to a balance yet. Please use the forum. That works great.

**Comment by abdullahtahiriyo** [comment]
Thanks @chennes for reviewing and testing. It really unloads a substantial part of the review process, as I am more focus on code reading and less on testing (if that will produce a good result, we will see ;) ).

**Pull request merged**